ref count write operations on IndexShard

brwe opened the pull request:
This commit adds a counter for IndexShard that keeps track of how many write operations are currently being processed. The counter is incremented whenever a write request is submitted in TransportShardReplicationOperationAction and decremented when it is finished.
The counter starts at one and can never go below 0. If it is a 0, all write operations that go through TransportShardReplicationOperationAction are blocked. This can be achieved by calling blockWriteOperations() on IndexShard and undone by calling unblockWriteOperations().

This is currently based on #10499. 

We can use this in IndexShard.close() to ensure that write operations will not succeed once the shard moved to closed. @bleskes can you take a look and see if this is what you meant?

bleskes commented on src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java:
```diff
@@ -130,10 +130,9 @@ protected ShardIterator shards(ClusterState clusterState, InternalRequest reques
 
     @Override
-    protected Tuple<BulkShardResponse, BulkShardRequest> shardOperationOnPrimary(ClusterState clusterState, PrimaryOperationRequest shardRequest) {
+    protected Tuple<BulkShardResponse, BulkShardRequest> shardOperationOnPrimary(ClusterState clusterState, PrimaryOperationRequest shardRequest, IndexShard indexShard) {
```
I wonder if we should also pass the IndexService since we already have it in the base class. Saves on the lookup.

bleskes commented on src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java:
```diff
@@ -237,7 +236,21 @@ public boolean isForceExecution() {
+                IndexShard indexShard = indicesService.indexServiceSafe(request.shardId.getIndex()).shardSafe(request.shardId.id());
+                boolean counterInc = false;
+                try {
+                    counterInc = indexShard.beforeWriteOperation();
```
this should be out of the try finally. and then we can avoid the counterInc boolean.

s1monw replied:
I think we can just move this all out of the try block and then we don't need to assign this to any value since int he finally block we simply decrement the counter. this would simplify this a lot I think

s1monw commented on src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -1290,4 +1308,109 @@ protected void operationProcessed() {
+     * @return true if the write operations counter could be incremented, false if the counter is at 0
+     */
+    public boolean beforeWriteOperation() {
+        synchronized (inFlightOperationsCounter) {
```
this is unclear to me why we need to sync here I also don't really understand the  entire class and why it needs the latch. I think waiting for 0 is a totally different problem an should be solved by a queue or so. There should be a base class counter that is very simple and we shouldn't mix all these usecases here

brwe commented:
Addressed all comments. Want to take another look?
I am unsure about the Integration test I added: SimpleIndexShardCounterIntegrationTests. It tests if replicas are failed in case the counter cannot be increased anymore. We should never get in this state so I do not know if we actually need to test this. If not I can also remove this check I guess: https://github.com/elastic/elasticsearch/pull/10610/files#diff-9669e07f0556311d187e534e321a0393R728

s1monw commented on src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -721,6 +725,9 @@ public void recover(Engine.RecoveryHandler recoveryHandler) throws EngineExcepti
     }
 
     public void failShard(String reason, Throwable e) {
+        if (engineClosed()) {
```
why do we need to check this here I think we should never check this and just call fail on it?

s1monw replied:
or if use engineUnsafe() and check for null

brwe replied:
I needed it for an Integration test I wrote, not sure if it makes sense though. See my comment above https://github.com/elastic/elasticsearch/pull/10610#issuecomment-93759161

s1monw commented on src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -746,18 +753,18 @@ public void close(String reason, boolean flushEngine) throws IOException {
                 }
                 changeState(IndexShardState.CLOSED, reason);
             } finally {
-                final Engine engine = this.currentEngineReference.getAndSet(null);
```
why are we not closing the engine here anymore? I think this is a bug

brwe replied:
Engine is closed when the counter goes to 0. IndexShard.close() only decrements it, the closing happens later in closeInternal(). Does that make sense?

s1monw replied:
I don't think we should do this. We are not ref-coutning the engine then we would do it on the engine. The engine should be close immediately on the close call in the IndexShard. This PR should only add the refCounting and the ability to wait until all references are returned. the purpose of this is to prevent crypting errors from the doc mapper etc. when indexing since we retry on EngineClosed exceptions the user won't see them.

bleskes replied:
Imho I think it's cleaner and better to wait on in flight operations before shutting down. For the same reason we wait on ongoing in engine operations before closing it. I never liked those logged EngineClosed exceptions, even if those are in trace. just my 2cents.

s1monw commented on src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -746,18 +753,18 @@ public void close(String reason, boolean flushEngine) throws IOException {
         }
     }
 
+    void waitForCloseCompletion() throws InterruptedException {
```
this should be public to be useful no? I also think we need javadocs what this does and maybe a timeout?

s1monw replied:
we have this twice now? there is antoher method below

s1monw commented on src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -999,6 +1012,23 @@ public void deleteShardState() throws IOException {
         MetaDataStateFormat.deleteMetaState(nodeEnv.shardPaths(shardId));
     }
 
+    public int getInFlightOperations() {
```
can this just be `getRefCount()` or maybe IndexShard implements RefCounted

s1monw commented on src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -1290,4 +1323,43 @@ protected void operationProcessed() {
+    private final AbstractRefCounted counter = new AbstractRefCounted("foo") {
+        @Override
+        protected void closeInternal() {
+            closeEngine();
```
hmm why are we closing the engine later? I think we should just close it on IndexShard#close

s1monw commented on src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -1290,4 +1323,43 @@ protected void operationProcessed() {
                 threadPool, indexingService, indexSettingsService, warmer, store, deletionPolicy, translog, mergePolicyProvider, mergeScheduler,
                 mapperAnalyzer, similarityService.similarity(), codecService, failedEngineListener, translogRecoveryPerformer);
     }
+
```
Is there a better name than "foo"

s1monw commented on src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -1290,4 +1323,43 @@ protected void operationProcessed() {
+    }
+
+
+    private final CountDownLatch latch = new CountDownLatch(1);
```
move this up to the other members?

s1monw commented on src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -1290,4 +1323,43 @@ protected void operationProcessed() {
+        }
+    };
+
+    private void closeEngine() {
```
this can go away?

brwe replied:
see my comment above https://github.com/elastic/elasticsearch/pull/10610#discussion_r28527958

s1monw commented on src/main/java/org/elasticsearch/index/engine/Engine.java:
```diff
@@ -1067,4 +1067,8 @@ public void close() throws IOException {
      */
     public abstract boolean hasUncommittedChanges();
+
+    public boolean isClosed() {
```
I don't think this is necessary

brwe replied:
I use it for testing. I can remove it but is there any harm in keeping?

s1monw replied:
I think we should not have this... just call a method and catch EngineCloseException if you wanna test it?

s1monw commented on src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -1290,4 +1323,43 @@ protected void operationProcessed() {
+    private final CountDownLatch latch = new CountDownLatch(1);
+
+
+    public void waitForEngineClose() throws InterruptedException {
```
I'd call this `waitForReferences` or so

s1monw commented on src/main/java/org/elasticsearch/index/IndexService.java:
```diff
@@ -395,6 +395,7 @@ private void closeShardInjector(String reason, ShardId sId, Injector shardInject
                     try {
                         final boolean flushEngine = deleted.get() == false && closed.get(); // only flush we are we closed (closed index or shutdown) and if we are not deleted
                         indexShard.close(reason, flushEngine);
+                        indexShard.waitForEngineClose();
```
can we have a unittest that this actually works?

s1monw commented on src/main/java/org/elasticsearch/common/util/concurrent/AbstractRefCounted.java:
```diff
@@ -28,7 +28,7 @@
  */
 public abstract class AbstractRefCounted implements RefCounted {
-    private final AtomicInteger refCount = new AtomicInteger(1);
+    protected final AtomicInteger refCount = new AtomicInteger(1);
```
I think this must be private use `refCount()` to get the value

s1monw commented:
I did another review of this

s1monw commented on src/test/java/org/elasticsearch/test/InternalTestCluster.java:
```diff
@@ -972,10 +974,31 @@ private synchronized void reset(boolean wipeData) throws IOException {
+     * Wipes any data that a test can leave behind: indices, templates and repositories
+     */
+    public void wipe() {
+        assertShardIndexCounter();
```
unneeded here?

s1monw commented on src/test/java/org/elasticsearch/test/InternalTestCluster.java:
```diff
@@ -972,10 +974,31 @@ private synchronized void reset(boolean wipeData) throws IOException {
+        super.wipe();
+    }
+
+    private void assertShardIndexCounter() {
```
this should be called in #assertAfterTest ?

brwe replied:
assertAfterTest() is called after wipe() (https://github.com/elastic/elasticsearch/blob/master/src/test/java/org/elasticsearch/test/ElasticsearchIntegrationTest.java#L657 ) so there would be indices left to check if I moved it there

s1monw commented on src/test/java/org/elasticsearch/index/shard/SimpleIndexShardCounterTests.java:
```diff
@@ -0,0 +1,209 @@
+import static org.hamcrest.Matchers.equalTo;
+
+
+public class SimpleIndexShardCounterTests extends ElasticsearchSingleNodeTest {
```
do we have  a  "ComplexIndexShardCounterTests" ? :) I think it can just go into `IndexShardTests`?

s1monw commented on src/test/java/org/elasticsearch/index/shard/SimpleIndexShardCounterIntegrationTests.java:
```diff
@@ -0,0 +1,192 @@
+/**
+ */
+@ElasticsearchIntegrationTest.ClusterScope(scope = ElasticsearchIntegrationTest.Scope.TEST, numDataNodes = 0)
+public class SimpleIndexShardCounterIntegrationTests extends ElasticsearchIntegrationTest {
```
I think this test is a bit over the top - the unit test below is good enough?

s1monw commented on src/test/java/org/elasticsearch/index/shard/SimpleIndexShardCounterTests.java:
```diff
@@ -0,0 +1,209 @@
+    }
+
+    @Test
+    public void testCloseShardCanCauseIndexShardClosedException() throws InterruptedException, ExecutionException, IOException {
```
these are all integration tests mostly with all the groovy stuff - I think we should not do this at all. I really don't like the sleeps here

s1monw commented on src/test/java/org/elasticsearch/index/shard/SimpleIndexShardCounterTests.java:
```diff
@@ -0,0 +1,209 @@
+
+public class SimpleIndexShardCounterTests extends ElasticsearchSingleNodeTest {
+
+    public void testDecRefClosesEngine() throws InterruptedException, ExecutionException {
```
can we have a very simple test that uses two thread to test the wait call?

s1monw commented:
I did another review...

>  am unsure about the Integration test I added: SimpleIndexShardCounterIntegrationTests. It tests if 

I think we don't need any integration test here really, we have enough integration tests that test this and your assertions are good too?

brwe commented:
I rebased on master now that #10749 has been pushed and replaced all integration tests by unit tests similar to the ones in #10749. I like the new test much more than the old ones but I had some trouble simulating an IndexShard and am unsure if the way I do it is a little too hacky. Please have another look!

s1monw commented on src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java:
```diff
@@ -947,4 +982,10 @@ public String concreteIndex() {
         }
     }
+
+    void decrementCounter(@Nullable IndexShard indexShard) {
```
I think we can simplify this a bit  and maybe have a class that does that ie this:

```Java

static class IndexShardReference {
  private final AtomicReference<IndexShard> ref = new AtomicReference();
  
  void setReference(IndexShard shard) {
     indexShard.incrementOperationCounter();
     assert ref.get() == null;
     ref.set(indexShard);
  } 

  void removeReference() {
    IndexShard shard = ref.get();
    try {
       shard. decrementOperationCounter();
    } finally {
      ref.set(null);
    } 
  }
}

```

this seems cleaner? or do I miss something

s1monw commented on src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -1302,4 +1314,24 @@ protected void operationProcessed() {
+
+    private final AbstractRefCounted counter = new AbstractRefCounted("in-flight-operations-counter") {
+        @Override
+        protected void closeInternal() {
```
maybe we can log here that we are done with this shard?

s1monw commented on src/test/java/org/elasticsearch/action/support/replication/ShardReplicationOperationTests.java:
```diff
@@ -566,7 +779,23 @@ protected boolean resolveIndex() {
+    * */
+    private IndexShard getIndexShard(ShardId shardId) {
+        try {
+            if (testIndexShard == null) {
```
can this be called concurrently? if so this needs synchronization otherwise we create it more than once?

s1monw commented on src/test/java/org/elasticsearch/action/support/replication/ShardReplicationOperationTests.java:
```diff
@@ -472,8 +522,172 @@ protected void runReplicateTest(IndexShardRoutingTable shardRoutingTable, int as
+        };
+        t.start();
+        // operation should be ongoing
+        awaitBusy(new Predicate<Object>() {
```
I don't think we should do a busy wait here. We might use a utility like this:

```Java
public static class Reference<T> {
        private T o;

        @SuppressForbidden(reason = "Uses wait / notifyAll")
        public synchronized T get() throws InterruptedException {
            while (o == null) {
                this.wait();
            }
           return o;
        }

        @SuppressForbidden(reason = "Uses wait / notifyAll")
        public synchronized void set(T o) {
            this.o = o;
            this.notifyAll();
        }

    }

```

brwe replied:
I tried and found that the await for not null was wrong to begin with. we should wait for counter to be at 2 instead. I think this would be a little more tricky to accomplish with a notify because the notify would then would have to happen on increment?

s1monw commented on src/test/java/org/elasticsearch/action/support/replication/ShardReplicationOperationTests.java:
```diff
@@ -582,5 +811,266 @@ static DiscoveryNode newNode(int nodeId) {
+
+    // from ShadowEngineTests
+    static protected Store createStore(ShardId shardId) throws IOException {
+        return createStore(new RAMDirectory(), shardId);
```
maybe use `newDirectory()` here instead of a dedicated ram directory

brwe replied:
if I do that then this is expected to be closed in the end but because this is only used for initializing the fake IndexShard it will never be closed

s1monw commented on src/test/java/org/elasticsearch/index/shard/IndexShardCounterTests.java:
```diff
@@ -0,0 +1,54 @@
+public class IndexShardCounterTests extends ElasticsearchSingleNodeTest {
+
+    @Test
+    public void testDeleteIndexDecreasesCounter() throws InterruptedException, ExecutionException, IOException {
```
can you move this into `IndexShardTests` it's also an instance of `ElasticsearchSingleNodeTest`

s1monw commented:
I did another review round looks very good. Left some cosmetic comments mainly

s1monw commented:
btw. @brwe I think you can trash most of the test code in here by simply using `RefCounted` here instead of hard coding IndexShard and IndicesService. I think this abstraction is totally ok and we should just test what we really need instead of mocking all this stuff everywhere.

brwe commented:
Thanks again for the review! Addressed all comments or answered. 

> I think you can trash most of the test code in here by simply using RefCounted here instead of hard 
> coding IndexShard and IndicesService.

Not sure what 'use' means here.
I now moved the ref counting in IndexShard to a separate class that I can use in tests and TransportShardReplicationOperationAction. @s1monw is this what you meant? I made it inherit from AbstractRefCounted also but am not sure this is needed or a good idea. I liked the method names incrementOperationsCounter better than incRef. I am not passionate though.

jpountz commented on src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java:
```diff
@@ -947,4 +983,23 @@ public String concreteIndex() {
+
+        void removeReference() {
+            RefCounted indexShardOperationCounter = ref.get();
+            ref.set(null);
```
`RefCounted indexShardOperationCounter = ref.getAndSet(null);` ?

brwe replied:
thanks!

jpountz commented on src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java:
```diff
@@ -947,4 +983,23 @@ public String concreteIndex() {
+        void setReference(RefCounted indexShardOperationCounter) {
+            indexShardOperationCounter.incRef();
+            assert ref.get() == null;
+            ref.set(indexShardOperationCounter);
```
```java
final RefCounter previous = ref.getAndSet(indexShardOperationCounter);
assert previous == null;
```

s1monw commented on src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java:
```diff
@@ -611,6 +627,10 @@ void retryBecauseUnavailable(ShardId shardId, String message) {
     }
 
+    protected RefCounted getIndexShardOperationsCounter(ShardId shardId) {
+        return indicesService.indexServiceSafe(shardId.index().getName()).shardSafe(shardId.id()).getOperationsCounter();
```
can we use multiple assignements here instead of this chaining?

s1monw commented on src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java:
```diff
@@ -947,4 +983,22 @@ public String concreteIndex() {
         }
     }
+
+    static class IndexShardReference {
```
here is a suggestion, why don't we make this class implement Closeable and reduce the entire thing into a single method we can override ie:

```
protected Closeable newShardReference(ShardId shardId) {
  // get the index shard
  return new IndexShardReference(indexShard);
}

// example usage
public void messageReceived(final ReplicaOperationRequest request, final TransportChannel channel) throws Exception {
             try(Closeable indexShardReference = newShardReference(request.shardId)) {
                 shardOperationOnReplica(request);
             } catch (Throwable t) {
                 failReplicaIfNeeded(request.shardId.getIndex(), request.shardId.id(), t);
                 throw t;
             }
             channel.sendResponse(TransportResponse.Empty.INSTANCE);
         }
```

that way we can keep you incrementOperationCounter methods on IndexShard but don't need to mock all the stuff?

brwe replied:
I made it a reference because we need to remember if we incremented in the PrimaryPhase (see https://github.com/elastic/elasticsearch/pull/10610#discussion_r29253762). I don't understand how making IndexShardReference implement Closeable would help here. Can you elaborate? 
In addition, if I I would override a method like newShardReference that only returns a Closeable than effectively the tests would test if TransportShardReplication work OK with whatever Closeable my test returns, not what IndexShard uses and that seems fishy to me. 
Or maybe I don't get it.

s1monw commented:
> incrementOperationsCounter better than incRef. I am not passionate though

I hear you! I added a comment with a suggestion to simplify it and bring `incrementOperationsCounter` back :) lemme know what you think

brwe commented:
Chatted with @s1monw and implemented all changes now. Want to have another look?

s1monw commented on src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java:
```diff
@@ -616,6 +630,13 @@ void retryBecauseUnavailable(ShardId shardId, String message) {
+        IndexService indexService = indicesService.indexServiceSafe(shardId.index().getName());
+        IndexShard indexShard = indexService.shardSafe(shardId.id());
+        return new IndexShardReference(indexShard);
+
```
extra newline

s1monw commented on src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java:
```diff
@@ -951,5 +988,25 @@ void concreteIndex(String concreteIndex) {
         public String concreteIndex() {
             return concreteIndex;
         }
+
```
extra newline

s1monw commented on src/test/java/org/elasticsearch/action/support/replication/ShardReplicationOperationTests.java:
```diff
@@ -472,8 +498,173 @@ protected void runReplicateTest(IndexShardRoutingTable shardRoutingTable, int as
+        assertIndexShardCounter(1);
     }
 
+    /*@Test
```
still needed?

s1monw commented on src/test/java/org/elasticsearch/test/TestCluster.java:
```diff
@@ -77,6 +77,13 @@ public void wipe() {
+     * Assertions that should run before the cluster is wiped should be called in this method
+     */
+    public void beforeIndexDeletion() {
+
```
extra newline

s1monw commented on src/test/java/org/elasticsearch/test/InternalTestCluster.java:
```diff
@@ -990,6 +971,23 @@ public synchronized void afterTest() throws IOException {
+            IndicesService indexServices = getInstance(IndicesService.class, nodeAndClient.name);
+            for (IndexService indexService : indexServices) {
+                for (IndexShard indexShard : indexService) {
+                    assertThat(indexShard.getOperationsCount(), equalTo(1));
```
what if this shard got relocated etc. then the counter is `0` for a short amount of time. I think we need to check this and if the counter is `0` here the shard must be closed so maybe we can check this too?

s1monw commented on src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java:
```diff
@@ -285,11 +279,19 @@ protected void responseWithFailure(Throwable t) {
         @Override
         protected void doRun() throws Exception {
-            shardOperationOnReplica(request.internalShardId, request);
+
```
extra newline

s1monw commented on src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java:
```diff
@@ -535,15 +546,18 @@ void performOnPrimary(final ShardRouting primary, final ShardIterator shardsIt)
                 if (retryPrimaryException(e)) {
                     logger.trace("had an error while performing operation on primary ({}), scheduling a retry.", e.getMessage());
+                    Releasables.close(indexShardReference);
+                    indexShardReference = null;
```
do we have to assign null here?

brwe replied:
yes, else when we retry and this fails again we will decrement in one of the finish methods although the counter was not incremented.

brwe replied:
we could alternatively also check if indexShardReference is null before we acquire a new reference I guess.

brwe commented:
pushed another commit

s1monw commented:
LGTM

brwe commented:
still have to backport to 1.x...

brwe commented:
pushed to 1.x
